polygon/heimdall/service.go

```diff
@@ -0,0 +1,158 @@
+	readerProvider := func() reader { /* TODO */ return nil }
+
+	tx := txProvider()
+	if tx == nil {
+		// TODO: implement and remove
+		logger.Warn("heimdall.Service txProvider is not implemented yet")
+		return nil
+	}
+	reader := readerProvider()
+	if reader == nil {
+		// TODO: implement and remove
+		logger.Warn("heimdall.Service readerProvider is not implemented yet")
+		return nil
+	}
+
+	blockNumToIdIndexFactory := func(ctx context.Context) (*RangeIndex, error) {
+		return NewRangeIndex(ctx, tmpDir, logger)
+	}
+
+	checkpointStore := newCheckpointStore(tx, reader, blockNumToIdIndexFactory)
+	milestoneStore := newMilestoneStore(tx, reader, blockNumToIdIndexFactory)
+	spanStore := newSpanStore(tx, reader, blockNumToIdIndexFactory)
+
+	client := NewHeimdallClient(heimdallUrl, logger)
+	scraper := NewScraper(
+		checkpointStore,
+		milestoneStore,
+		spanStore,
+		client,
+		1*time.Second,
+		logger,
+	)
+
+	return &service{
+		scraper: scraper,
+
+		checkpointStore: checkpointStore,
+		milestoneStore:  milestoneStore,
+		spanStore:       spanStore,
+	}
+}
+
+func (s *service) FetchLatestSpan(ctx context.Context) (*Span, error) {
+	s.scraper.Synchronize(ctx)
+	entity, err := s.spanStore.GetLastEntity(ctx)
+	return downcastSpanEntity(entity), err
+}
+
+func castEntityToWaypoint(entity Entity) Waypoint {
+	return entity.(Waypoint)
+}
+
+func (s *service) FetchCheckpointsFromBlock(ctx context.Context, startBlock uint64) (Waypoints, error) {
+	s.scraper.Synchronize(ctx)
+	entities, err := s.checkpointStore.RangeFromBlockNum(ctx, startBlock)
+	return libcommon.SliceMap(entities, castEntityToWaypoint), err
+}
+
+func (s *service) FetchMilestonesFromBlock(ctx context.Context, startBlock uint64) (Waypoints, error) {
+	s.scraper.Synchronize(ctx)
+	entities, err := s.milestoneStore.RangeFromBlockNum(ctx, startBlock)
+	return libcommon.SliceMap(entities, castEntityToWaypoint), err
+}
+
+func sliceTakeLast[T any](s []T, count int) []T {
+	length := len(s)
+	if length > count {
+		return s[length-count:]
+	}
+	return s
+}
+
+// TODO: this limit is a temporary solution to avoid piping thousands of events
+// during the first sync. Let's discuss alternatives. Hopefully we can remove this limit.
+const maxEntityEvents = 5
+
+func (s *service) OnMilestoneEvent(callback func(*Milestone)) polygoncommon.UnregisterFunc {
+	return s.scraper.RegisterMilestoneObserver(func(entities []*Milestone) {
+		for _, entity := range sliceTakeLast(entities, maxEntityEvents) {
+			callback(entity)
+		}
+	})
+}
+
+func (s *service) OnSpanEvent(callback func(*Span)) polygoncommon.UnregisterFunc {
+	return s.scraper.RegisterSpanObserver(func(entities []*Span) {
+		for _, entity := range sliceTakeLast(entities, maxEntityEvents) {
+			callback(entity)
+		}
+	})
+}
+
+func (s *service) Run(ctx context.Context) error {
+	defer s.checkpointStore.Close()
+	defer s.milestoneStore.Close()
+	defer s.spanStore.Close()
+
+	prepareStoresGroup, _ := errgroup.WithContext(ctx)
```

Comment:
lost `ctx` param. 
child `ctx` will cancel when any goroutine return error/panic

Reply:
:+1: good spot. Initially the Prepare method didn't accept a ctx, and I've passed a wrong context after a refactoring.

Reply:
adressed here: https://github.com/ledgerwatch/erigon/pull/10337